File: core/src/main/java/io/github/shaksternano/borgar/core/media/MediaReaderFactory.java
package io.github.shaksternano.borgar.core.media;

import io.github.shaksternano.borgar.core.media.reader.MediaReader;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;

public interface MediaReaderFactory<T> {

    MediaReader<T> createReader(File media, String format) throws IOException;

    MediaReader<T> createReader(InputStream media, String format) throws IOException;
}

File: core/src/main/kotlin/io/github/shaksternano/borgar/core/media/MediaReaderFactory.kt
package io.github.shaksternano.borgar.core.media

import io.github.shaksternano.borgar.core.io.DataSource
import io.github.shaksternano.borgar.core.media.reader.AudioReader
import io.github.shaksternano.borgar.core.media.reader.ImageReader
import io.github.shaksternano.borgar.core.media.reader.MediaReader

interface MediaReaderFactory<T : VideoFrame<*>> {

    val supportedFormats: Set<String>

    suspend fun create(input: DataSource): MediaReader<T>
}

interface ImageReaderFactory : MediaReaderFactory<ImageFrame> {
    override suspend fun create(input: DataSource): ImageReader
}

interface AudioReaderFactory : MediaReaderFactory<AudioFrame> {
    override suspend fun create(input: DataSource): AudioReader
}
